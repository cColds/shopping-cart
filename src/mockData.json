{
  "data": {
    "products": {
      "edges": [
        {
          "node": {
            "id": "gid://shopify/Product/7982853619734",
            "title": "Slides",
            "description": "Simple, minimal and comfortable, these slides feature a classic design in the perfect shade of iron. Whether you're just lounging around the house or running errands, these slides will offer all-day comfort.",
            "featuredImage": {
              "id": "gid://shopify/ProductImage/39774608687126",
              "url": "https://cdn.shopify.com/s/files/1/0688/1755/1382/products/slides.jpg?v=1675447358"
            },
            "variants": {
              "edges": [
                {
                  "node": {
                    "price": {
                      "amount": "25.0",
                      "currencyCode": "CAD"
                    }
                  }
                }
              ]
            }
          }
        },
        {
          "node": {
            "id": "gid://shopify/Product/7982856273942",
            "title": "Sweatpants",
            "description": "Soft and comfortable sweatpants in stylish shades. They are perfect for lounging with their cozy stretch fabric that offers just the right amount of warmth. Enjoy the ultimate relaxation experience!",
            "featuredImage": {
              "id": "gid://shopify/ProductImage/39774602002454",
              "url": "https://cdn.shopify.com/s/files/1/0688/1755/1382/products/GreenSweatpants01.jpg?v=1675455387"
            },
            "variants": {
              "edges": [
                {
                  "node": {
                    "price": {
                      "amount": "35.0",
                      "currencyCode": "CAD"
                    }
                  }
                }
              ]
            }
          }
        }
      ]
    }
  }
}
