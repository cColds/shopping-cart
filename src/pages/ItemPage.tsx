import { useEffect } from "react";
import { useParams } from "react-router-dom";
import { FaCirclePlus, FaCircleMinus } from "react-icons/fa6";
import Item from "../interfaces/Item";

function ItemPage({
  items,
  onAddToCartClick,
  onDecrementClick,
  onIncrementClick,
  onQuantityInputChange,
  quantity,
  resetQuantity,
  currentItem,
  setCurrentItem,
}: {
  items: Item[];
  onAddToCartClick: () => void;
  onDecrementClick: () => void;
  onIncrementClick: () => void;
  onQuantityInputChange: (quantity: number) => void;
  quantity: number;
  resetQuantity: () => void;
  currentItem: Item | null;
  setCurrentItem: (updateItem: Item | null) => void;
}) {
  const { itemId } = useParams();

  useEffect(() => {
    resetQuantity();
  }, [itemId]);

  useEffect(() => {
    const targetItem =
      items.find(
        ({ node: { title } }) =>
          title.replace(/\s+/g, "-").toLowerCase() ===
          itemId?.replace(/\s+/g, "-").toLowerCase()
      ) || null;
    setCurrentItem(targetItem);
  }, [items, itemId]);

  if (currentItem == null) return null;
  const price = parseFloat(
    currentItem.node.variants.edges[0].node.price.amount
  );

  return (
    <div className="flex flex-col flex-wrap gap-5 p-8">
      <div className="flex flex-wrap justify-center gap-5">
        <img
          src={currentItem.node.featuredImage.url}
          alt={currentItem.node.title}
          width="300px"
          height="300px"
          className="rounded-lg"
          draggable="false"
        />

        <div className="flex h-min flex-col gap-5 rounded-lg border-[1px] border-slate-200 p-2.5">
          <div>
            <h1 className="text-xl">{currentItem.node.title}</h1>
            <p className="font-poppins-reg text-lg">${price}</p>
          </div>
          <div className="flex gap-1.5">
            <button
              onClick={(e) => {
                if ((e.target as HTMLButtonElement).disabled) return;

                onDecrementClick();
              }}
              aria-label="decrease quantity"
              disabled={quantity === 1}
            >
              <FaCircleMinus className="text-2xl" />
            </button>
            <input
              onChange={(e) => {
                let quantity = e.target.valueAsNumber;
                if (quantity > 1000) quantity = 1000;
                else if (quantity < 1) quantity = 1;

                onQuantityInputChange(quantity);
              }}
              aria-label="item quantity"
              type="number"
              value={quantity}
              className="h-[30px] w-full rounded-lg border-[1px] border-slate-200 text-center"
            />
            <button
              onClick={onIncrementClick}
              aria-label="increase quantity"
              className="text-2xl"
              disabled={quantity >= 1000}
            >
              <FaCirclePlus />
            </button>
          </div>
          <button
            onClick={onAddToCartClick}
            className="w-full overflow-hidden truncate rounded-lg bg-primary-color p-2 text-center text-sm font-bold text-white"
          >
            Add to cart
          </button>
        </div>
      </div>
      <div className="min-w-full max-w-min border-t-[1px] border-slate-200">
        <h3 className="my-1">Description</h3>
        <p>{currentItem.node.description}</p>
      </div>
    </div>
  );
}

export default ItemPage;
