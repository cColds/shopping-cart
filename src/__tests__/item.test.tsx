import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import { MemoryRouter, Routes, Route } from "react-router";
import ItemPage from "../pages/ItemPage";
import mockData from "../mockData.json";

let itemCount: number;
let quantity: number;

beforeEach(() => {
  itemCount = 0;
  quantity = 1;
  vi.clearAllMocks();
});

const mockHandleAddToCart = vi.fn(() => {
  itemCount += quantity;
});

const mockHandleDecrement = vi.fn(() => {
  quantity -= 1;
});

const mockHandleIncrement = vi.fn(() => {
  quantity += 1;
});

const mockHandleQuantity = vi.fn();

const mockResetQuantity = vi.fn();

it("should increment item count when clicking add to cart", async () => {
  const user = userEvent.setup();
  const items = mockData.data.products.edges;

  render(
    <MemoryRouter initialEntries={["/store/slides"]}>
      <Routes>
        <Route
          path="/store/:itemId"
          element={
            <ItemPage
              items={items}
              onAddToCartClick={mockHandleAddToCart}
              onDecrementClick={mockHandleDecrement}
              onIncrementClick={mockHandleIncrement}
              onQuantityInputChange={mockHandleQuantity}
              quantity={quantity}
              resetQuantity={mockResetQuantity}
            />
          }
        />
      </Routes>
    </MemoryRouter>
  );

  await user.click(screen.getByRole("button", { name: /add to cart/i }));
  await user.click(screen.getByRole("button", { name: /add to cart/i }));
  await user.click(screen.getByRole("button", { name: /add to cart/i }));

  expect(mockHandleAddToCart).toBeCalledTimes(3);
  expect(itemCount).toBe(3);
});

it("should increment quantity by 1", async () => {
  const user = userEvent.setup();
  const items = mockData.data.products.edges;

  render(
    <MemoryRouter initialEntries={["/store/slides"]}>
      <Routes>
        <Route
          path="/store/:itemId"
          element={
            <ItemPage
              items={items}
              onAddToCartClick={mockHandleAddToCart}
              onDecrementClick={mockHandleDecrement}
              onIncrementClick={mockHandleIncrement}
              onQuantityInputChange={mockHandleQuantity}
              quantity={quantity}
              resetQuantity={mockResetQuantity}
            />
          }
        />
      </Routes>
    </MemoryRouter>
  );

  await user.click(screen.getByRole("button", { name: /increase quantity/i }));

  expect(mockHandleIncrement).toBeCalledTimes(1);
  expect(quantity).toBe(2);
});

it("should decrement quantity by 1", async () => {
  const user = userEvent.setup();
  const items = mockData.data.products.edges;
  quantity = 3;
  render(
    <MemoryRouter initialEntries={["/store/slides"]}>
      <Routes>
        <Route
          path="/store/:itemId"
          element={
            <ItemPage
              items={items}
              onAddToCartClick={mockHandleAddToCart}
              onDecrementClick={mockHandleDecrement}
              onIncrementClick={mockHandleIncrement}
              onQuantityInputChange={mockHandleQuantity}
              quantity={quantity}
              resetQuantity={mockResetQuantity}
            />
          }
        />
      </Routes>
    </MemoryRouter>
  );

  await user.click(screen.getByRole("button", { name: /increase quantity/i }));
  await user.click(screen.getByRole("button", { name: /increase quantity/i }));
  await user.click(screen.getByRole("button", { name: /decrease quantity/i }));
  expect(mockHandleDecrement).toBeCalledTimes(1);
  expect(quantity).toBe(4);
});

it("should not decrement quantity by 1 if it's already 1", async () => {
  const user = userEvent.setup();
  const items = mockData.data.products.edges;

  render(
    <MemoryRouter initialEntries={["/store/slides"]}>
      <Routes>
        <Route
          path="/store/:itemId"
          element={
            <ItemPage
              items={items}
              onAddToCartClick={mockHandleAddToCart}
              onDecrementClick={mockHandleDecrement}
              onIncrementClick={mockHandleIncrement}
              onQuantityInputChange={mockHandleQuantity}
              quantity={quantity}
              resetQuantity={mockResetQuantity}
            />
          }
        />
      </Routes>
    </MemoryRouter>
  );

  await user.click(screen.getByRole("button", { name: /decrease quantity/i }));

  expect(mockHandleDecrement).toBeCalledTimes(0);
  expect(quantity).toBe(1);
});

it("should set input quantity to 65", async () => {
  const user = userEvent.setup();
  const items = mockData.data.products.edges;

  render(
    <MemoryRouter initialEntries={["/store/slides"]}>
      <Routes>
        <Route
          path="/store/:itemId"
          element={
            <ItemPage
              items={items}
              onAddToCartClick={mockHandleAddToCart}
              onDecrementClick={mockHandleDecrement}
              onIncrementClick={mockHandleIncrement}
              onQuantityInputChange={mockHandleQuantity}
              quantity={quantity}
              resetQuantity={mockResetQuantity}
            />
          }
        />
      </Routes>
    </MemoryRouter>
  );
  const quantityInput = screen.getByRole("spinbutton", {
    name: /item quantity/i,
  });
  await user.type(quantityInput, "65");
  expect(mockHandleQuantity).toHaveBeenCalledTimes(2);
});
