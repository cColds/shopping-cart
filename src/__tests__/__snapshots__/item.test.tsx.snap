// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render item page and match snapshot 1`] = `
<div>
  <div
    class="flex flex-col flex-wrap gap-5 p-8"
  >
    <div
      class="flex flex-wrap justify-center gap-5"
    >
      <img
        alt="Slides"
        class="rounded-lg"
        draggable="false"
        height="300px"
        src="https://cdn.shopify.com/s/files/1/0688/1755/1382/products/slides.jpg?v=1675447358"
        width="300px"
      />
      <div
        class="flex h-min flex-col gap-5 rounded-lg border-[1px] border-slate-200 p-2.5"
      >
        <div>
          <h1
            class="text-xl"
          >
            Slides
          </h1>
          <p
            class="font-poppins-reg text-lg"
          >
            $
            25
          </p>
        </div>
        <div
          class="flex gap-1.5"
        >
          <button
            aria-label="decrease quantity"
            disabled=""
          >
            <svg
              class="text-2xl"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM184 232H328c13.3 0 24 10.7 24 24s-10.7 24-24 24H184c-13.3 0-24-10.7-24-24s10.7-24 24-24z"
              />
            </svg>
          </button>
          <input
            aria-label="item quantity"
            class="h-[30px] w-full rounded-lg border-[1px] border-slate-200 text-center"
            type="number"
            value="1"
          />
          <button
            aria-label="increase quantity"
            class="text-2xl"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM232 344V280H168c-13.3 0-24-10.7-24-24s10.7-24 24-24h64V168c0-13.3 10.7-24 24-24s24 10.7 24 24v64h64c13.3 0 24 10.7 24 24s-10.7 24-24 24H280v64c0 13.3-10.7 24-24 24s-24-10.7-24-24z"
              />
            </svg>
          </button>
        </div>
        <button
          class="w-full overflow-hidden truncate rounded-lg bg-primary-color p-2 text-center text-sm font-bold text-white"
        >
          Add to cart
        </button>
      </div>
    </div>
    <div
      class="min-w-full max-w-min border-t-[1px] border-slate-200"
    >
      <h3
        class="my-1"
      >
        Description
      </h3>
      <p>
        Simple, minimal and comfortable, these slides feature a classic design in the perfect shade of iron. Whether you're just lounging around the house or running errands, these slides will offer all-day comfort.
      </p>
    </div>
  </div>
</div>
`;
